### PR TITLE
Atf master+linaro warp7 squash v3

### PR DESCRIPTION
**Pull request by bryanodonoghue**
This is version #3 of the WaRP7 TF-A port pull request.

Taking the input from @danh-arm I have squashed down the patches further, breaking the link between our git history entirely in favour of contiguous chunks of functionality.
There are 40 patches here, which is 50 patches less than the last PR and a full 110 less than the initial pass.
I could potentially squash down the imx: clock, imx7: clock, and io_mux: set further but, I feel there would be a loss of fidelity in that.

Changelog:

1. Reworked patches to remove the chronology of our git history
2. Squashed down 50 patches in the process
3. Made a directory tweak adding plat/imx/imx7/common/imx7_clock.c a change I had meant to make but had forgotten about until now
4. Added Acked-by: Anson Huang <Anson.Huang@nxp.com> as indicated for the first patch
5. Issuing PR against integration branch as requested.
@niej @rmsilva @omnium21 @antonio-nino-diaz-arm

### CODE REVIEW

**Comment by antonio-nino-diaz-arm** [comment]
One initial comment, more will follow. Your C console driver works right now because you are using the deprecated driver (``drivers/console/aarch64/deprecated_console.S``) that simply ends every function by calling the ``core`` console functions. However, the multi console API won't work with this driver because the functions in ``drivers/console/aarch64/multi_console.S``. The functions in this driver use x14 as a temporary iterator register, which is a scratch register according to the AAPCS, so it doesn't need to be preserved by the compiler. It may still work even after switching to the multi console driver, but it would be because of pure luck. Only x18 - x29 are preserved.

I see two options:
1. Rewrite this driver in assembly like the others. This would allow it to be reused by the plat console instead of having two drivers like you have now.
2. Fix the multi console driver so that it uses the correct registers according to the convention.

We will deprecate the old driver at some point, and I see the value in having a C driver for the console, so I think that the correct thing to do is 2.

**Comment by antonio-nino-diaz-arm** [comment]
The files in plat/imx/common should have more specific names, they are too generic. Prefix them by `imx_`.

``plat/imx/imx7/include/hab_arch.h`` and ``plat/imx/imx7/include/imx-regs.h`` should also be renamed.

File: plat/imx/common/snvs.h [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+	uint32_t lptdcr;
+	uint32_t lpsr;
+	uint32_t lpsrtcmr;
+	uint32_t lpsrtclr;
+	uint32_t lptar;
+	uint32_t lpsmcmr;
+	uint32_t lpsmclr;
+	uint32_t lppgdr;
+	uint32_t lpgpr0_alias;
+	uint8_t  lpzmkr[32];
```
[the remark this answers is not in the document]
Comment (bryanodonoghue) [reply plat/imx/common/snvs.h]: Yes it should.

File: plat/imx/common/caam.h [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+#define DECORR_OFFSET	0x9C
+#define RES3_SIZE DECORR_OFFSET - RTIC_MID_OFFSET - \
+	(sizeof(struct caam_rtic_mid) * CAAM_NUM_RTIC)
+
+#define DECO_MID_OFFSET	0xA0
+#define DAR_OFFSET	0x120
+#define RES4_SIZE DAR_OFFSET - DECO_MID_OFFSET - \
+	(sizeof(struct caam_deco) * CAAM_NUM_DECO)
+
+struct caam_ctrl {
```
[the remark this answers is not in the document]
Comment (bryanodonoghue) [reply plat/imx/common/caam.h]: In theory no. In practice yes
Theoretically the sizeof the reserved space is aligned.
In practice I don't think that is guaranteed.

**Comment by bryanodonoghue** [comment]
Sorry @antonio-nino-diaz-arm 

Do you really mean to prefix files inside of plat/imx/common/ with an extra imx_ ? The names are highly specific
AIPS ->

**Pull request closed without merging**